sklearn/utils/__init__.py

```diff
@@ -178,22 +178,63 @@ def axis0_safe_slice(X, mask, len_mask):
     return np.zeros(shape=(0, X.shape[1]))
 
 
-def safe_indexing(X, indices):
+def safe_indexing(X, indices, axis=0):
+    """Return rows, items or columns of X using indices.
+
+    Parameters
+    ----------
+    X : array-like, sparse-matrix, list, pandas.DataFrame, pandas.Series
+        Data from which to sample rows, items or columns.
+    indices : see below
+        When ``axis=0``, indices need to be an array of integer.
```

Comment:
Array-like I presume?

I'm a bit uncomfortable with the disparity between different axis handling, particularly with support for things like callables, and lack of support for Boolean mask on samples.

Reply:
I would have like just merging both functions at first. I agree that we could support the same indexing in both axis (apart of string for axis=0).

Shall we make it in another PR or in this PR. It is already a big diff.

Reply:
Fixing the consistency in another PR will also allow me to work in parallel for the PDP/dataframe support.

Reply:
Get rid of callables and I'll approve that

Reply:
Checking the tests and the code, I don't see anywhere that we actually detect or test for callable.